### ISSUE TITLE
Emulated mouse and keyboard keep waking computer instantly after it goes to sleep

### ISSUE THREAD

**Issue by fennectech** [Valkirie/HandheldCompanion#754]
**Device manufacturer**
- [ ] AYANEO
- [ ] AYN
- [ ] AOKZOE
- [ X] ASUS ROG ALLY
- [ ] GPD
- [ ] ONEXPLAYER
- [ ] VALVE

**Device model**
ROG Ally

**Handheld Companinion Version**
0.16.2.5

**Describe the bug**
Emulated mouse and keyboard hid devices instantly wake up the device after sleep

**Expected behavior**
Device should stay asleep

I have found a workaround by disabling “allow this device to wake the system” in device manager on both your emulated “hid compliant mouse” and “Hid compliant keyboard”

**Comment by CasperH2O** [comment]
@fennectech good catch, this needs to be improved indeed. I recall this being reported before, but can't find a ticket. Perhaps we can make it so that these emulated devices get these settings when HC creates them.

**Comment by fennectech** [comment]
Does windows even let you mess with that stuff programmatically?     You may have to modify the registry directly!

**Comment by CasperH2O** [comment]
To my knowledge we already have some code in HC to adjust HID device properties. As for this specific case, here's what ChatGPT came up with:

> ChatGPT, with the C# programming language how can I change a HID device setting that I normally find in device called "allow this device to wake the system".

> In C#, you can change a HID (Human Interface Device) device setting using the Windows Device Management APIs. Specifically, you would need to modify the device properties related to power management to enable or disable the "allow this device to wake the system" setting.
> 
> Here's a general outline of how you might achieve this:
> 
>     Use the DeviceInformation class to locate the specific HID device you want to modify. You can enumerate devices and filter them based on their device class (HID) and other criteria.
> 
>     Retrieve the device's device property key for the power management settings using the DevicePropertyKey class.
> 
>     Use the DeviceProperty class to access the power management settings of the device.
> 
>     Modify the power management settings to enable or disable the "allow this device to wake the system" option.
> 
>     Use the DeviceManagement APIs to apply the changes to the device.
> 
> Here's a code snippet illustrating the concept (note that this is a simplified example and might need further refinement based on your specific use case):

```
using System;
using System.Runtime.InteropServices;
using Windows.Devices.Enumeration;
using Windows.Devices.HumanInterfaceDevice;

class Program
{
    static async Task Main(string[] args)
    {
        string targetDeviceId = "your_device_id_here";

        DeviceInformationCollection devices = await DeviceInformation.FindAllAsync(HidDevice.GetDeviceSelector());
        
        DeviceInformation targetDevice = devices.FirstOrDefault(device => device.Id == targetDeviceId);
        
        if (targetDevice != null)
        {
            HidDevice hidDevice = await HidDevice.FromIdAsync(targetDevice.Id, FileAccessMode.ReadWrite);

            if (hidDevice != null)
            {
                Guid powerManagementGuid = new Guid("your_power_management_guid_here");

                DevicePropertyKey powerManagementKey = new DevicePropertyKey(powerManagementGuid, yourPropertyId);

                DeviceProperty powerManagementProperty = await hidDevice.GetDevicePropertyAsync(powerManagementKey);
                
                // Modify the power management property to enable or disable the wake setting
                // powerManagementProperty.Value = newValue;

                await powerManagementProperty.SaveToDeviceAsync();

                hidDevice.Dispose();
            }
            else
            {
                Console.WriteLine("Failed to open HID device.");
            }
        }
        else
        {
            Console.WriteLine("Device not found.");
        }
    }
}

```

**Comment by fennectech** [comment]
ChatGPT is freaking awesome

**Comment by CasperH2O** [comment]
Tried to reproduce, cannot get the device to fall asleep. 

Switching off the screen after one minute works with both HC running or not. Sleep mode does not seem to work after the timeout, regardless if HC is running or not and with all HID, mouse and keyboard devices having the "Allow this device to wake the computer" setting unticked. Triggering sleep mode from the Start menu does work and the device goes to sleep properly.

Will investigate further.
